packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,415 @@
+
+    const availableNamespaces = fromEntries(
+      await Promise.all(
+        filteredSnaps.map(async (snap) => [
+          snap.id,
+          await this.snapToNamespaces(snap),
+        ]),
+      ),
+    );
+
+    // The magical matching algorithm specified in SIP-2
+    const namespaceToSnaps = findMatchingKeyringSnaps(
+      connection.requiredNamespaces,
+      availableNamespaces,
+    );
+
+    const permissions = await this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    );
+
+    const approvedNamespacesAndSnaps = Object.entries(namespaceToSnaps).reduce(
+      (acc, [namespace, snapIds]) => {
+        // If snap already is approved for use, solve conflict by using that snap
+        const approvedSnap = snapIds.find((snapId) => {
+          return (
+            permissions &&
+            hasProperty(permissions, getSnapPermissionName(snapId))
+          );
+        });
+
+        if (approvedSnap) {
+          return { ...acc, [namespace]: [approvedSnap] };
+        }
+        return acc;
+      },
+      {},
+    );
+
+    const hasConflicts = Object.keys(namespaceToSnaps).some(
+      (namespace) => !hasProperty(approvedNamespacesAndSnaps, namespace),
+    );
+
+    const filteredNamespacesAndSnaps = hasConflicts
+      ? namespaceToSnaps
+      : approvedNamespacesAndSnaps;
+
+    const possibleAccounts = await this.namespacesToAccounts(
+      origin,
+      filteredNamespacesAndSnaps,
+      connection.requiredNamespaces,
+    );
+
+    // If current configuration doesn't solve request, we need to show a prompt.
+    const resolvedAccounts = hasConflicts
+      ? await this.resolveConflicts(origin, possibleAccounts)
+      : fromEntries(
+          Object.entries(possibleAccounts).map(
+            ([namespace, snapAndAccounts]) => [
+              namespace,
+              snapAndAccounts[0] ?? null,
+            ],
+          ),
+        );
+
+    // TODO: In the future, use another permission here to not give full permission after handshake.
+    // Instead we should use a less scary permission.
+    const approvedPermissions = Object.values(resolvedAccounts).reduce(
+      (acc, cur) => {
+        if (cur) {
+          return {
+            ...acc,
+            [getSnapPermissionName(cur.snapId)]: {},
+          };
+        }
+        return acc;
+      },
+      {},
+    );
+
+    await this.messagingSystem.call('PermissionController:grantPermissions', {
+      approvedPermissions,
+      subject: { origin },
+    });
+
+    const providedNamespaces: Record<NamespaceId, SessionNamespace> =
+      fromEntries(
+        Object.entries(connection.requiredNamespaces)
+          .filter(([namespaceId]) => resolvedAccounts[namespaceId]?.accounts)
+          .map(([namespaceId, namespace]) => [
+            namespaceId,
+            {
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
Generally as a defensive programming thing, prefer doing assertions with `assert` rather than ignoring lint errors.

Reply:
I understand, it just feels very unnecessary here given the filter.

Thoughts @Mrtenz ?

Reply:
I agree with Olaf. The overhead of doing assert is minimal, and it's safer than type casting or using non-null assertions.

Reply:
Rewrote to be use `reduce` instead, that seems easier to type and read.